### ISSUE TITLE
Recover from a failed command distribution

### ISSUE THREAD

**Issue by remcowesterhoud** [camunda/camunda#24300]
**Description**

Command distributions are used to send command from partition 1 to the other partitions. This way we enable the other partitions to process changes in resources that should be shared between all partitions, such as processes.

There is no guarantee that this distribution succeeds. As there likely is network traffic between brokers there's a chance this will fail. For this reason a retry mechanism has been introduces. The receiving partitions must acknowledge the processing of a command back to partition 1. Only once this acknowledgement is received will partition 1 stop retrying the distribution.

With the merge of [this PR](https://github.com/camunda/camunda/pull/24099) deployments and deletions are now distributed in an ordered manner. Only once partition 1 has received an acknowledgement of the previous distribution, it will continue with the next distribution.

All this becomes problematic once distribution during on processing the distributed command. When this happens partition 1 will never receive an acknowledgement. It will infinitely keep retrying to distribute the command. Worse is that this infinite retry will block all of the other distributions that are in the same queue. So if 1 deployment distribution fails, it becomes impossible to distribute any other deployments or deletions, as they are all on the same queue and blocked by the distribution that is stuck.

We must find a way to recover from these scenarios. We've already[ seen a bug ](https://github.com/camunda/camunda/issues/14055)where the deployment distribution fails to process as expected on receiving partitions. If this same bug where to occur in combination with ordered distribution, we have no way to recover the cluster.

**Comment by npepinpe** [comment]
Out of curiosity, why is this marked as feature instead of bug? 🤔 From a user point of view, this is effectively a bug, right?

> All this becomes problematic once distribution during on processing the distributed command.

I don't understand this sentence, but I assume you mean if distribution fails, i.e. there's a bug? So this issue is a feature (not bug) because the bug would be whatever caused distribution to fail, right? 😅  And you just want a way to bypass it or something.

**Comment by remcowesterhoud** [comment]
@npepinpe 
> Out of curiosity, why is this marked as feature instead of bug? 🤔 From a user point of view, this is effectively a bug, right?

I struggled to decide what it was to be honest. From my perspecive we want to implement a new feature to modify data on a specific partition. Or maybe we want to allow distributing a user-provided command so we can restore the data in that way to unblock the stuck distibutions. Looking at it from that perspective I'd consider it a feature.

The root-cause for the failed command distribution would be the bug. An easy way to recover from it is a feature imo.

> I don't understand this sentence, but I assume you mean if distribution fails, i.e. there's a bug?

That's because that sentence is a trainwreck 😆 I've re-written it. It's what you said. If processing the distributed command fails all distributions get stuck while we infinitely retry this distribution.

**Comment by remcowesterhoud** [comment]
@eppdot Can I raise some awareness of this issue? Yesterday @berkaycanbc @korthout and me have been working together on a cluster that had problems which were likely caused due to command distributions. We found it is difficult to rootcause and we can't treat any symptoms without making some code changes. I see this as a high risk as command distributions are getting used more often.

I don't think we have to implement all the improvements at once. Even focussing on the smaller ones would already help us out in the future.

**Comment by korthout** [comment]
Thanks for writing this up @remcowesterhoud 👏 

I believe all proposals make sense. How about we make some sub-issues out of the suggested changes, so we can prioritize them individually, and keep this as a way to track it all.

>The root-cause for the failed command distribution would be the bug. An easy way to recover from it (without a new release) is a feature imo.

I align with @npepinpe on this one, as command distribution is an implementation detail. That the issues associated to it requires additional management tools to overcome is bug fixing. I'd expect us to backport monitoring, pausing distribution retrying, and configurable backoffs. I'm not entirely certain about querying and canceling individual distributions, although these could be backported marked as `alpha` as they might be needed to recover from bugs.

**Comment by npepinpe** [comment]
I think that could be a good onboarding topic for our new team member, however I need to see if we can find capacity for it. I agree that this is pretty important (mostly because of the consequences of not having it).

**Comment by remcowesterhoud** [comment]
> I believe all proposals make sense. How about we make some sub-issues out of the suggested changes, so we can prioritize them individually, and keep this as a way to track it all.

Done

**Comment by eppdot** [comment]
I fully agree that it's a super crucial to monitor and control command distribution. We also stumbled across this in Vienna, where we fixed a bug regarding message [duplication](https://github.com/camunda/camunda/pull/27525) and more debugging capabilities around command distribution would have been super useful during the connected support case.

I am going to investigate the first issue about monitoring #29492 a bit and see if I can make some progress.

**Comment by npepinpe** [comment]
Thanks, appreciated! For the rest, I think this is a good onboarding for Carlo for dynamic scaling, but ofc if you want to tackle them before he's back, you're welcome 😁
